crates/assets/src/asset.rs

```diff
@@ -0,0 +1,901 @@
+		let mut split = input.splitn(2, |c| *c == NAMESPACE_REFERENCE_SEPARATOR);
+		let (namespace, reference) = (split.next(), split.next());
+
+		// Split the remaining reference to extract the identifier, if present
+		let (reference, identifier) = if let Some(r) = reference {
+			let mut split = r.splitn(2, |c| *c == REFERENCE_IDENTIFIER_SEPARATOR);
+			// Split the reference further, if present
+			(split.next(), split.next())
+		} else {
+			// Return the old reference, which is None if we are at this point
+			(reference, None)
+		};
+
+		(namespace, reference, identifier)
+	}
+
+	/// A Slip44 asset reference.
+	/// It is a modification of the [CAIP-20 spec](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-20.md)
+	/// according to the rules defined in the Asset DID method specification.
+	#[non_exhaustive]
+	#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, RuntimeDebug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+	pub struct Slip44Reference(pub U256);
+
+	impl Slip44Reference {
+		/// Parse a UTF8-encoded decimal Slip44 asset reference, failing if the
+		/// input string is not valid.
+		pub(crate) fn from_utf8_encoded<I>(input: I) -> Result<Self, AssetIdError>
+		where
+			I: AsRef<[u8]> + Into<Vec<u8>>,
+		{
+			let input = input.as_ref();
+			check_reference_length_bounds(input)?;
+
+			let decoded = str::from_utf8(input).map_err(|_| ReferenceError::InvalidFormat)?;
+			let parsed = U256::from_dec_str(decoded).map_err(|_| ReferenceError::InvalidFormat)?;
+			// Unchecked since we already checked for maximum length and hence maximum value
+			Ok(Self(parsed))
+		}
+	}
+
+	impl TryFrom<U256> for Slip44Reference {
+		type Error = AssetIdError;
+
+		fn try_from(value: U256) -> Result<Self, Self::Error> {
+			// Max value for 64-digit decimal values (used for Slip44 references so far).
+			// TODO: This could be enforced at compilation time once constraints on generics
+			// will be available.
+			if value
+				<= U256::from_str_radix("9999999999999999999999999999999999999999999999999999999999999999", 10)
+					.expect("Casting the maximum value for a Slip44 reference into a U256 should never fail.")
+			{
+				Ok(Self(value))
+			} else {
+				Err(ReferenceError::TooLong.into())
+			}
+		}
+	}
+
+	impl From<u128> for Slip44Reference {
+		fn from(value: u128) -> Self {
+			Self(value.into())
+		}
+	}
+
+	impl Display for Slip44Reference {
+		fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+			write!(f, "{}", self.0)
+		}
+	}
+
+	/// An asset reference that is identifiable only by an EVM smart contract
+	/// (e.g., a fungible token). It is a modification of the [CAIP-21 spec](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-21.md)
+	/// according to the rules defined in the Asset DID method specification.
+	#[non_exhaustive]
+	#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, RuntimeDebug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+	pub struct EvmSmartContractFungibleReference(pub [u8; 20]);
```

Comment:
You use the `non_exhaustive` because you want the `EvmSmartContractFungibleReference` to only be created by a `new` call function. Put you don't want to make the inner part private because then you would need to write getters?

Reply:
Yes, correct. You mentioned something about implementing Default?

Reply:
I think if you implement default you are still able to construct the type even though you have the `#[non_exhaustive]`.

I think the main issue with the `#[non_exhaustive]` is, that nobody will know why it's there. If you know nothing about the code and you read that. What would you think it does? "EvmSmartContractFungibleReference might get additional fields in the future." would be the first guess.

would it not be ok to have: 

```rust
pub struct Slip44Reference(pub(crate) U256);

impl AsRef<U256> for Slip44Reference {
	fn as_ref(&self) -> &U256 {
		&self.0
	}
}
```